File: LC49_GroupAnagrams/Hash/hash.py
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

# Input: strs = ["eat","tea","tan","ate","nat","bat"]
# Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
# Input: strs = [""]
# Output: [[""]]
# Input: strs = ["a"]
# Output: [["a"]]

from typing import List


def groupAnagrams(strs: List[str]) -> List[List[str]]:
    # Hash
    # Time Complexity: O(nk)
    # Space Complexity: O(nk)
    if len(strs) <= 1:
        return [strs]
    
    result = {}

    # count the number of each letter in the word
    for str in strs:
        count_table = [0]*26 # [0,0,0...,0]
        for char in str:
            # ord('a'): 97
            # ord('b'): 98
            # ord('c'): 99
            count_table[ord(char)-ord('a')] += 1 # add one in each index correspond to the letter
        # if count_table is the same, then same word but different sequence of letter
        key = tuple(count_table) # tuple is hashable while list is not
        result[key] = result.get(key,[]) + [str]
    return result.values()

# # # # # # # # # # # # # # # # # # # # Test Cases # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
print("test 1: ")
print(groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

print("test 2: ")
print(groupAnagrams([""]))

print("test 3: ")
print(groupAnagrams(["a"]))

File: LC49_GroupAnagrams/Hash/Hash.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Hash{
    public List<List<String>> groupAnagrams(String[] strs){
        HashMap<String,ArrayList<String>> result = new HashMap<>();
        for(String str : strs){
            char[] char_array = str.toCharArray();
            int[] count_table = new int[26];
            for(char c : char_array){
                count_table[c-'a']++;
            }
            // make array as string in order to be hashable
            // [1,12,3] -> "#1#12#3"
            StringBuilder sb = new StringBuilder();
            for(int block : count_table){
                sb.append("#");
                sb.append(block);
            }
            String key = sb.toString();
            if(!result.containsKey(key)) result.put(key, new ArrayList<>());
            result.get(key).add(str);
        }
        return new ArrayList(result.values());
    }
}